wp-content/plugins/tapestry/controller/class.user-controller.php

```diff
@@ -0,0 +1,141 @@
+<?php
+// TODO Change exceptions to using an ERROR class
+/**
+ * Add/update/retrieve User progress
+ * 
+ */
+class TapestryUserController {
+
+    /** 
+     * Update User's video progress for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @nodeId the current node being viewed
+     * @param type @progressValue is how much the video was viewed, value should be between >= 0 and <= 1
+    */
+    public function updateProgress($postId, $nodeId, $progressValue) {
+        $userId = apply_filters('determine_current_user', false);
+        $this->checkUserAndPostId($userId, $postId);
+
+        if ($progressValue !== NULL) {
+            $progressValue = json_decode($progressValue);
+        }
+
+        // Value should be between 0 and 1
+        if ($progressValue < 0 || $progressValue > 1) {
+            throw new Exception('Invalid progress value');
+        }
+
+        $this->updateUserProgress($userId, $postId, $nodeId, $progressValue);
+    }
+
+    /** 
+     * Get User's video progress for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @nodeIdArr is a list of ids currently in the tapestry
+    */
+    public function getProgress($postId, $nodeIdArr) {
```

Comment:
I think it's better if you get the array of nodes in the backend, rather than as a parameter here. Unless you have a reason why you need to have that passed in?

Reply:
This would require me to get Andrew's new tapestrycontroller code. I would be blocked again.

Reply:
Yes, we use get_post_meta(). But since we split our pull requests, his tapestry class is not in my PR, and I do not have the functions to add nodes to the db to test that functionality. My current code is using the ids from tapestry.json to update the progress value in the user_meta. 

I could take his latest code in his PR, since his PR is not merged theres a possibility that the code is changed.

Reply:
Please use the `getTapestryNodeIds` function you added in the tapestry controller.

Reply:
I mean why are you passing in `$nodeIdArr` instead of just retrieving the node ids using the `getTapestryNodeIds` function?